apps/src/lib/proto/types.rs

```diff
@@ -0,0 +1,457 @@
+use std::collections::hash_map::DefaultHasher;
+use std::convert::TryFrom;
+use std::hash::{Hash, Hasher};
+
+use prost::Message;
+use prost_types::Timestamp;
+use thiserror::Error;
+
+use super::generated::{services, types};
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("Error decoding a transaction from bytes: {0}")]
+    TxDecodingError(prost::DecodeError),
+    #[error("Error decoding an IntentGossipMessage from bytes: {0}")]
+    IntentDecodingError(prost::DecodeError),
+    #[error("Error decoding an DkgGossipMessage from bytes: {0}")]
+    DkgDecodingError(prost::DecodeError),
+    #[error("Intent is empty")]
+    NoIntentError,
+    #[error("Dkg is empty")]
+    NoDkgError,
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Tx {
+    pub code: Vec<u8>,
+    pub data: Option<Vec<u8>>,
+    pub timestamp: Timestamp,
+}
+
+impl TryFrom<&[u8]> for Tx {
+    type Error = Error;
+
+    fn try_from(tx_bytes: &[u8]) -> Result<Self> {
+        let tx = types::Tx::decode(tx_bytes).map_err(Error::TxDecodingError)?;
+        let timestamp = match tx.timestamp {
+            Some(t) => t,
+            None => std::time::SystemTime::now().into(),
```

Comment:
Because the `timestamp` is required in a valid transaction, I think we should return error when the `timestamp` is missing instead